src/c-writer.cc

```diff
@@ -547,35 +577,57 @@ std::string_view StripLeadingDollar(std::string_view name) {
   return name;
 }
 
-std::string CWriter::DefineImportName(const std::string& name,
-                                      std::string_view module,
-                                      std::string_view field_name) {
+void CWriter::DefineImportName(const Import* import,
+                               std::string_view module,
+                               std::string_view field_name) {
   std::string mangled = MangleName(module) + MangleName(field_name);
-  import_syms_.insert(name);
-  global_syms_.insert(mangled);
-  global_sym_map_.insert(SymbolMap::value_type(name, mangled));
-  return mangled;
-}
+  std::string name;
+  ModuleFieldType type;
+
+  switch (import->kind()) {
+    case ExternalKind::Func:
+      type = ModuleFieldType::Func;
+      name = cast<FuncImport>(import)->func.name;
+      break;
+    case ExternalKind::Tag:
+      type = ModuleFieldType::Tag;
+      name = cast<TagImport>(import)->tag.name;
+      break;
+    case ExternalKind::Global:
+      type = ModuleFieldType::Global;
+      name = cast<GlobalImport>(import)->global.name;
+      break;
+    case ExternalKind::Memory:
+      type = ModuleFieldType::Memory;
+      name = cast<MemoryImport>(import)->memory.name;
+      break;
+    case ExternalKind::Table:
+      type = ModuleFieldType::Table;
+      name = cast<TableImport>(import)->table.name;
+      break;
+  }
 
-std::string CWriter::DefineImportInstanceName(const std::string& name,
-                                              std::string_view module,
-                                              std::string_view field_name) {
-  std::string mangled = MangleName(module) + MangleName(field_name);
   import_syms_.insert(name);
+  import_module_sym_map_.emplace(name, import->module_name);
   global_syms_.insert(mangled);
-  global_sym_map_.insert(SymbolMap::value_type(name, mangled));
-  return "(*" + mangled + ")";
+  bool success =
+      global_sym_map_.emplace(name + MangleField(type), mangled).second;
+  assert(success);
 }
 
+template <ModuleFieldType T>
 std::string CWriter::DefineGlobalScopeName(std::string_view name) {
```

Comment:
Does this need to be a template?  Can it not just take ModuleFieldType as a second argument?

Reply:
Sure. I guess I was thinking that the value is always known at compile-time, but it can be an argument just as easily. Done.